File: app/src/main/java/dev/chu/architecture/java/p87/MainActivityComponent.java
package dev.chu.architecture.java.p87;

import dagger.BindsInstance;
import dagger.Subcomponent;
import dev.chu.architecture.etc.ActivityScope;

@Subcomponent(modules = MainActivityModule.class)
@ActivityScope
public interface MainActivityComponent {

    MainFragmentComponent.Builder mainFragmentComponentBuilder();

    void inject(MainActivity activity);

    @Subcomponent.Builder
    interface Builder {
        Builder setModule(MainActivityModule module);
        @BindsInstance
        Builder setActivity(MainActivity activity);
        MainActivityComponent build();
    }

}

File: app/src/main/java/dev/chu/architecture/kotlin/p87/MainActivityComponent.kt
package dev.chu.architecture.kotlin.p87

import dagger.BindsInstance
import dagger.Subcomponent
import dev.chu.architecture.etc.ActivityScope

@Subcomponent(modules = [MainActivityModule::class])
@ActivityScope
interface MainActivityComponent {
    fun mainFragmentFactory(): MainFragmentComponent.Factory
    fun inject(activity: MainActivity)

    // 액티비티 인스턴스는
    @Subcomponent.Factory
    interface Factory {
        fun create(
            module: MainActivityModule,
            @BindsInstance activity: MainActivity
        ): MainActivityComponent
    }
}
